lib/services/local/access_list.go

```diff
@@ -259,67 +278,87 @@ func (a *AccessListService) DeleteAllAccessListMembers(ctx context.Context) erro
 func (a *AccessListService) UpsertAccessListWithMembers(ctx context.Context, accessList *accesslist.AccessList, membersIn []*accesslist.AccessListMember) (*accesslist.AccessList, []*accesslist.AccessListMember, error) {
 	// Double the lock TTL to account for the time it takes to upsert the members.
-	err := a.service.RunWhileLocked(ctx, lockName(accessList.GetName()), 2*accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
-		// Create a map of the members from the request for easier lookup.
-		membersMap := make(map[string]*accesslist.AccessListMember)
-
-		// Convert the members slice to a map for easier lookup.
-		for _, member := range membersIn {
-			membersMap[member.GetName()] = member
-		}
-
-		var (
-			members      []*accesslist.AccessListMember
-			membersToken string
-			err          error
-		)
-
-		for {
-			// List all members for the access list.
-			members, membersToken, err = a.memberService.WithPrefix(accessList.GetName()).ListResources(ctx, 0 /* default size */, membersToken)
-			if err != nil {
-				return trace.Wrap(err)
+	upsertWithLockFn := func() error {
+		return a.service.RunWhileLocked(ctx, lockName(accessList.GetName()), 2*accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
+			// Create a map of the members from the request for easier lookup.
+			membersMap := make(map[string]*accesslist.AccessListMember)
+
+			// Convert the members slice to a map for easier lookup.
+			for _, member := range membersIn {
+				membersMap[member.GetName()] = member
 			}
 
-			for _, member := range members {
-				// If the member is not in the members map (request), delete it.
-				if _, ok := membersMap[member.GetName()]; !ok {
-					err = a.memberService.WithPrefix(accessList.GetName()).DeleteResource(ctx, member.GetName())
-					if err != nil {
-						return trace.Wrap(err)
-					}
-				} else {
-					// Compare members and update if necessary.
-					if !cmp.Equal(member, membersMap[member.GetName()]) {
-						// Update the member.
-						err = a.memberService.WithPrefix(accessList.GetName()).UpsertResource(ctx, membersMap[member.GetName()])
+			var (
+				members      []*accesslist.AccessListMember
+				membersToken string
+				err          error
+			)
+
+			for {
+				// List all members for the access list.
+				members, membersToken, err = a.memberService.WithPrefix(accessList.GetName()).ListResources(ctx, 0 /* default size */, membersToken)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				for _, member := range members {
+					// If the member is not in the members map (request), delete it.
+					if _, ok := membersMap[member.GetName()]; !ok {
+						err = a.memberService.WithPrefix(accessList.GetName()).DeleteResource(ctx, member.GetName())
 						if err != nil {
 							return trace.Wrap(err)
 						}
+					} else {
+						// Compare members and update if necessary.
+						if !cmp.Equal(member, membersMap[member.GetName()]) {
```

Comment:
Are we sure here that all fields should be compared? From what I see `AccessListMember` contains `IneligibleStatus string` which may or may not be set as we compute this field on the fly, also inside the `ResourceHeader.Metadata` we keep `Revision string` which is an internal identifier and I don't think we should update the resource when it changes.

Reply:
unrelated to my PR (as i'm just placing limits)

i'm unsure of the other fields, but for `ineligibleStatus`, we strip this field [FromProto](https://github.com/gravitational/teleport/blob/1bd30f6d5ed391f3e6eec5916ec0741aa99ea521/api/types/accesslist/convert/v1/member.go#L47)